File: tests/tainting_rules/java/try_return.java
// https://github.com/returntocorp/semgrep/issues/4091
// We want to test that the try-block is properly connected to the
// catch-block despite the try-block execution ends in a `return`.
public class Main {
    public void test1() {
        String str = "safe";
        try {
            return could_throw();
        }
        catch (Throwable t) {
            str = source();
        }
        //ruleid: test
        sink(str);
    }

    public void test2() {
        int cannot_throw = 42;
        String str = "safe";
        try {
            return cannot_throw;
        }
        // vvv dead code
        catch (Throwable t) {
            str = source();
        }
        //OK:
        sink(str);
    }
}

File: tests/tainting_rules/dart/try_return.dart
// Translated from java/try_return.java

class Main {
  // These simulate the Java helper methods
  String? couldThrow() {
    // may throw
    return null;
  }

  String source() {
    return "tainted";
  }

  void sink(String s) {
    print("sink: $s");
  }

  // -------------------------
  // test1
  // -------------------------
  String? test1() {
    String str = "safe";
    try {
      // Java: return could_throw();
      // Dart: method returns String?, so this is valid
      return couldThrow();
    } catch (e) {
      str = source();
    }

    // ruleid: test
    sink(str);
    return null;
  }

  // -------------------------
  // test2
  // -------------------------
  int? test2() {
    int cannotThrow = 42;
    String str = "safe";

    try {
      // Java: return cannot_throw;
      // Dart: method returns int?, so this is valid
      return cannotThrow;
    }
    // vvv dead code (preserved exactly)
    catch (e) {
      str = source();
    }

    // OK:
    sink(str);
    return null;
  }
}
